force-app/main/default/classes/SOQL.cls

```diff
@@ -1,361 +1,585 @@
+    // Create a set of email domains from the Leads
+    // Add the email domain to the set
+    for (Lead ld : leadEmailLst) {
+      emailDomains.add('%' + ld.Email.split('@').get(1));
     }
 
-    /**
-     * Question 23
-     * Clones and Inserts Opportunity Line Items from an original Opportunity to a list of cloned Opportunities.
-     * Adds a randomized number between 1 to 3 to the quantity of each line item.
-     * Example: cloneLineItems('001XXXXXXXXXXXX', [list of opportunity ids]) should clone the line items from the Opportunity with the Id to each Opportunity in the list
-     * @param Original Opportunity Id
-     * @param List of cloned Opportunity Ids
-     */
-    public static void cloneLineItems(Id origOppId, List<Id> cloneOppIds) {
-        // Query the original Opportunity Line Items
-
-        // Create a list to hold the new line items
-
+    // Query the Accounts
+    return [SELECT Id, Website FROM Account WHERE Website LIKE :emailDomains];
+  }
+
+  /**
+   * Question 23
+   * Clones and Inserts Opportunity Line Items from an original Opportunity to a list of cloned Opportunities.
+   * Adds a randomized number between 1 to 3 to the quantity of each line item.
+   * Example: cloneLineItems('001XXXXXXXXXXXX', [list of opportunity ids]) should clone the line items
+   * from the Opportunity with the Id to each Opportunity in the list
+   * @param Original Opportunity Id
+   * @param List of cloned Opportunity Ids
+   */
+  public static void cloneLineItems(Id origOppId, List<Id> cloneOppIds) {
+    // Create a list to hold the new line items
+    List<OpportunityLineItem> newLineItemLst = new List<OpportunityLineItem>();
+
+    if (!String.isBlank(origOppId) && !cloneOppIds.isEmpty()) {
+      // Query the original Opportunity Line Items
+      List<OpportunityLineItem> oppLineItemLst = [
+        SELECT
+          Id,
+          Description,
+          ListPrice,
+          Name,
+          Product2Id,
+          PricebookEntryId,
+          ProductCode,
+          Quantity,
+          ServiceDate,
+          SortOrder,
+          UnitPrice
+        FROM OpportunityLineItem
+        WHERE OpportunityId = :origOppId
+      ];
+      // Loop through each cloned Opportunity Id
+      for (Id oppId : cloneOppIds) {
         // Generate a random number between 1 and 3
-
-        // Loop through each cloned Opportunity Id
-
-            // Loop through each original line item
-
-                // Clone the original line item and add the random number to the quantity
-
-        // Insert the new line items
+        Integer randNum = Math.mod(Integer.valueOf(Math.random() * 10), 4);
+        // Loop through each original line item
+        for (OpportunityLineItem oli : oppLineitemLst) {
+          // Clone the original line item and add the random number to the quantity
+          OpportunityLineItem oppLineItem = oli.clone(
+            false,
+            true,
+            false,
+            false
+          );
```

Comment:
For practice try clone this manually by setting the values in the new oppLineItem instead of using he clone method.